### PR TITLE
Alerting: Customise OK notification priorities for Pushover notifier

### LINKED ISSUE #27190: Pushover: support for different priority for OK
If a Pushover alert of the priority EMERGENCY is sent I am woken up even if my iPhone is on silent or do not disturb. That is the way it is meant to be.  However, the OK notification for the same alert is still sent out using EMERGENCY priority, waking me up again :/

@Hofls implemented the option to choose a different sound for the OK alert a while ago, which is great. Now with the new EMERGENCY priority enabled on iOS by Apple for Pushover it would be great if we could choose a separate PRIORITY (i.e. default) for the OK alert too.

### PR DESCRIPTION
**Pull request by acaire**
Not sure if removing the `ShowWhen` properties was the best idea, but felt cleaner than either duplicating the `Expire` and `Retry` values for UI consumption and/or trying to work out how to turn [Is](https://github.com/grafana/grafana/blob/1cae5c9a2179f41b9e6e67b2847501e06dacacdc/pkg/services/alerting/notifier.go#L83), into checking an array instead.

Closes #27190

### CODE REVIEW

File: pkg/services/alerting/notifiers/pushover.go [hunk 1]
```diff
@@ -193,7 +193,8 @@ func NewPushoverNotifier(model *models.AlertNotification) (alerting.Notifier, er
 	userKey := model.DecryptedValue("userKey", model.Settings.Get("userKey").MustString())
 	APIToken := model.DecryptedValue("apiToken", model.Settings.Get("apiToken").MustString())
 	device := model.Settings.Get("device").MustString()
-	priority, _ := strconv.Atoi(model.Settings.Get("priority").MustString())
+	alertingPriority, _ := strconv.Atoi(model.Settings.Get("priority").MustString())
```
Comment (wbrowne) [inline pkg/services/alerting/notifiers/pushover.go:196]: I think you could make the field name `alertingPriority`, and use this opportunity to make the `sound` field consistent too IE `sound` -> `alertingSound`

**Review by wbrowne — COMMENTED** [review]
Looks great - thanks for the contribution 😄  Just one small nit while you're in there changing things!

File: pkg/services/alerting/notifiers/pushover.go (hunk 1 above)
Comment (acaire) [reply pkg/services/alerting/notifiers/pushover.go:196]: Thanks for the review @wbrowne 😄 ! This seems to break my existing config as I gather it's expecting these values straight from the database (Unless there's a migration pattern I'm unaware of)?

Comment (acaire) [reply pkg/services/alerting/notifiers/pushover.go:196]:
```diff
diff --git a/pkg/services/alerting/notifiers/pushover.go b/pkg/services/alerting/notifiers/pushover.go
index 59f225b568..46267de717 100644
--- a/pkg/services/alerting/notifiers/pushover.go
+++ b/pkg/services/alerting/notifiers/pushover.go
@@ -150,7 +150,7 @@ func init() {
 				Label:         "Alerting priority",
 				Element:       alerting.ElementTypeSelect,
 				SelectOptions: priorityOptions,
-				PropertyName:  "priority",
+				PropertyName:  "alertingPriority",
 			},
 			{
 				Label:         "OK priority",
@@ -176,7 +176,7 @@ func init() {
 				Label:         "Alerting sound",
 				Element:       alerting.ElementTypeSelect,
 				SelectOptions: soundOptions,
-				PropertyName:  "sound",
+				PropertyName:  "alertingSound",
 			},
 			{
 				Label:         "OK sound",
@@ -193,11 +193,11 @@ func NewPushoverNotifier(model *models.AlertNotification) (alerting.Notifier, er
 	userKey := model.DecryptedValue("userKey", model.Settings.Get("userKey").MustString())
 	APIToken := model.DecryptedValue("apiToken", model.Settings.Get("apiToken").MustString())
 	device := model.Settings.Get("device").MustString()
-	alertingPriority, _ := strconv.Atoi(model.Settings.Get("priority").MustString())
+	alertingPriority, _ := strconv.Atoi(model.Settings.Get("alertingPriority").MustString())
 	okPriority, _ := strconv.Atoi(model.Settings.Get("okPriority").MustString())
 	retry, _ := strconv.Atoi(model.Settings.Get("retry").MustString())
 	expire, _ := strconv.Atoi(model.Settings.Get("expire").MustString())
-	alertingSound := model.Settings.Get("sound").MustString()
+	alertingSound := model.Settings.Get("alertingSound").MustString()
 	okSound := model.Settings.Get("okSound").MustString()
 	uploadImage := model.Settings.Get("uploadImage").MustBool(true)
```

Comment (wbrowne) [reply pkg/services/alerting/notifiers/pushover.go:196]: Actually yeah in retrospect, it's not worth the trouble right now. Apologies!

Comment (acaire) [reply pkg/services/alerting/notifiers/pushover.go:196]: No problems! Thanks for confirming 😃

Comment (aknuds1) [inline pkg/services/alerting/notifiers/pushover.go:196]:
```suggestion
	alertingPriority, err := strconv.Atoi(model.Settings.Get("priority").MustString())
	if err != nil {
		return fmt.Errorf("failed to convert alerting priority to integer: %w", err)
	}
```

File: pkg/services/alerting/notifiers/pushover.go [hunk 2]
```diff
@@ -193,7 +193,8 @@ func NewPushoverNotifier(model *models.AlertNotification) (alerting.Notifier, er
...
+	okPriority, _ := strconv.Atoi(model.Settings.Get("okPriority").MustString())
```
Comment (aknuds1) [inline pkg/services/alerting/notifiers/pushover.go:197]:
```suggestion
	okPriority, err := strconv.Atoi(model.Settings.Get("okPriority").MustString())
	if err != nil {
		return fmt.Errorf("failed to convert OK priority to integer: %w", err)
	}
```

File: pkg/services/alerting/notifiers/pushover.go [hunk 3]
```diff
@@ -207,34 +208,36 @@ func NewPushoverNotifier(model *models.AlertNotification) (alerting.Notifier, er
...
-		Device:        device,
-		AlertingSound: alertingSound,
-		OkSound:       okSound,
-		Upload:        uploadImage,
-		log:           log.New("alerting.notifier.pushover"),
+		NotifierBase:     NewNotifierBase(model),
+		UserKey:          userKey,
+		APIToken:         APIToken,
+		AlertingPriority: alertingPriority,
+		OkPriority:       okPriority,
```
Comment (aknuds1) [inline pkg/services/alerting/notifiers/pushover.go:215]: Standard Go style dictates that initialisms be written in consistent case:
```suggestion
		OKPriority:       okPriority,
```

File: pkg/services/alerting/notifiers/pushover.go [hunk 4]
```diff
@@ -207,34 +208,36 @@ func NewPushoverNotifier(model *models.AlertNotification) (alerting.Notifier, er
...
-	Expire        int
-	Device        string
-	AlertingSound string
-	OkSound       string
-	Upload        bool
-	log           log.Logger
+	UserKey          string
+	APIToken         string
+	AlertingPriority int
+	OkPriority       int
```
Comment (aknuds1) [inline pkg/services/alerting/notifiers/pushover.go:233]:
```suggestion
	OKPriority       int
```

File: pkg/services/alerting/notifiers/pushover_test.go [hunk 5]
```diff
@@ -54,7 +55,8 @@ func TestPushoverNotifier(t *testing.T) {
 				So(pushoverNotifier.Type, ShouldEqual, "pushover")
 				So(pushoverNotifier.APIToken, ShouldEqual, "4SrUFQL4A5V5TQ1z5Pg9nxHXPXSTve")
 				So(pushoverNotifier.UserKey, ShouldEqual, "tzNZYf36y0ohWwXo4XoUrB61rz1A4o")
-				So(pushoverNotifier.Priority, ShouldEqual, 1)
+				So(pushoverNotifier.AlertingPriority, ShouldEqual, 1)
+				So(pushoverNotifier.OkPriority, ShouldEqual, 2)
```
Comment (aknuds1) [inline pkg/services/alerting/notifiers/pushover_test.go:59]:
```suggestion
				So(pushoverNotifier.OKPriority, ShouldEqual, 2)
```

**Review by aknuds1 — CHANGES_REQUESTED** [review]
Generally looks good, but I see some missing error checks (the old code also fails to check errors, but that's an anti-pattern).

**Comment by acaire** [comment]
Thanks for your review @aknuds1, I wasn't aware of initialisms - I accepted all of your suggestions, would you like me to squash?

**Comment by aknuds1** [comment]
> Thanks for your review @aknuds1, I wasn't aware of initialisms - I accepted all of your suggestions, would you like me to squash?

Thanks! No need to squash, we squash merge anyway.

**Comment by aknuds1** [comment]
> Thanks for your review @aknuds1, I wasn't aware of initialisms - I accepted all of your suggestions, would you like me to squash?

Thanks! No need to squash, we squash merge anyway.

File: pkg/services/alerting/notifiers/pushover.go [hunk 6]
```diff
@@ -322,12 +331,16 @@ func (pn *PushoverNotifier) genPushoverBody(evalContext *alerting.EvalContext, m
 	}
 
 	// Add priority
-	err = w.WriteField("priority", strconv.Itoa(pn.Priority))
+	priority := pn.AlertingPriority
+	if evalContext.Rule.State == models.AlertStateOK {
+		priority = pn.OkPriority
```
Comment (aknuds1) [inline pkg/services/alerting/notifiers/pushover.go:336]:
```suggestion
		priority = pn.OKPriority
```

**Review by aknuds1 — APPROVED** [review]
LGTM, but see my suggestion to fix a compile error.

File: pkg/services/alerting/notifiers/pushover.go (hunk 6 above)
Comment (acaire) [reply pkg/services/alerting/notifiers/pushover.go:336]: Ahh appreciate it thanks, I did notice that but was going to wait until I was back near my computer.

File: pkg/services/alerting/notifiers/pushover.go [hunk 7]
```diff
@@ -207,34 +214,36 @@ func NewPushoverNotifier(model *models.AlertNotification) (alerting.Notifier, er
...
-	log           log.Logger
+	UserKey          string
+	APIToken         string
+	AlertingPriority int
+	OKPriority       int
+	Retry            int
+	Expire           int
+	Device           string
+	AlertingSound    string
+	OkSound          string
```
Comment (acaire) [inline pkg/services/alerting/notifiers/pushover.go:244]: @aknuds1 would you like me to fix this one while i'm here?

Comment (aknuds1) [reply pkg/services/alerting/notifiers/pushover.go:244]: Sure, yes please.

**Comment by aknuds1** [comment]
There's another [build failure](https://drone.grafana.net/grafana/grafana/9908/1/6) BTW.

**Comment by acaire** [comment]
@aknuds1 hopefully that's got it 🤞

**Review by aknuds1 — APPROVED** [review]
LGTM

@wbrowne @papagian Have either of you tested this PR?

**Comment by aknuds1** [comment]
I noticed that my suggestions for error handling hadn't been committed, only resolved. Assuming that was by accident, I committed them myself. That caused a trivial build failure though which I fixed and pushed to your branch.

**Comment by acaire** [comment]
> I noticed that my suggestions for error handling hadn't been committed, only resolved. Assuming that was by accident, I committed them myself. That caused a trivial build failure though which I fixed and pushed to your branch.

Yeah apologies i could've sworn i pushed with your changes incorporated, they're definitely in my local 🤔

**Comment by aknuds1** [comment]
I'll wait with merging it until I know that someone on our team has tested it.

**Comment by aknuds1** [comment]
Looks as if @papagian will test the PR.

**Review by papagian — CHANGES_REQUESTED** [review]
Thanks for you contribution! I have requested some changes.

In addition to these, I think it would be could to set a default value for `retry` and `expire` (in case of alerting emergency priority) because if I've selected `Emergency` I get a validation error:
`DBUG[01-19|21:57:28] Webhook failed                           logger=notifications url=https://api.pushover.net/1/messages.json statuscode="400 Bad Request" body="{\"retry\":\"is too small, must be at least 30 seconds\",\"errors\":[\"retry is invalid\"],\"status\":0,\"request\":\"617141b9-4226-4652-9470-4502c84eef8c\"}"`

**Review by marefr — CHANGES_REQUESTED** [review]
Can you please update the provisioning documentation here as well: https://github.com/grafana/grafana/blob/master/docs/sources/administration/provisioning.md#alert-notification-pushover

**Review by papagian — APPROVED** [review]
I have tested it and works fine; thank you for your contribution! I have pushed some documentation updates.

File: docs/sources/alerting/notifications.md [hunk 8]
```diff
@@ -130,6 +130,23 @@ Move any existing rules using `custom_details.myMetric` to `custom_details.queri
 This behavior will become the default in a future version of Grafana.
 
 > Using `dedup_key` tag will override Grafana generated `dedup_key` with a custom key.
+
+### Pushover
+
+To set up Pushover, all you have to do is to provide a User key and an API token. You can follow [What is Pushover and how do I use it](https://support.pushover.net/i7-what-is-pushover-and-how-do-i-use-it) on how to generate them.
```
Comment (oddlittlebird) [inline docs/sources/alerting/notifications.md:136]:
```suggestion
To set up Pushover, you must provide a user key and an API token. Refer to [What is Pushover and how do I use it](https://support.pushover.net/i7-what-is-pushover-and-how-do-i-use-it) for instructions on how to generate them.
```

File: docs/sources/alerting/notifications.md [hunk 9]
```diff
@@ -130,6 +130,23 @@ Move any existing rules using `custom_details.myMetric` to `custom_details.queri
...
+
+
+Setting | Description
+---------- | -----------
+API Token | Application token
+User key(s) | A comma-separated list of User keys
```
Comment (oddlittlebird) [inline docs/sources/alerting/notifications.md:142]:
```suggestion
User key(s) | A comma-separated list of user keys
```

File: docs/sources/alerting/notifications.md [hunk 10]
```diff
@@ -130,6 +130,23 @@ Move any existing rules using `custom_details.myMetric` to `custom_details.queri
...
+Device(s) | A comma-separated list of devices
+Priority | The priority alerting nottifications are sent
+OK priority | The priority OK nottifications are sent; if not set OK notifications will be sent with the priotiry set for alerting notifications
```
Comment (oddlittlebird) [inline docs/sources/alerting/notifications.md:145]:
```suggestion
OK priority | The priority OK notifications are sent; if not set, then OK notifications are sent with the priority set for alerting notifications 
```

File: docs/sources/alerting/notifications.md [hunk 11]
```diff
@@ -130,6 +130,23 @@ Move any existing rules using `custom_details.myMetric` to `custom_details.queri
...
+
+
+Setting | Description
+---------- | -----------
+API Token | Application token
+User key(s) | A comma-separated list of User keys
+Device(s) | A comma-separated list of devices
+Priority | The priority alerting nottifications are sent
+OK priority | The priority OK nottifications are sent; if not set OK notifications will be sent with the priotiry set for alerting notifications 
+Retry | How often (in seconds) the Pushover servers will send the same notification to the user. (minimun 30 seconds)
```
Comment (oddlittlebird) [inline docs/sources/alerting/notifications.md:146]:
```suggestion
Retry | How often (in seconds) the Pushover servers send the same notification to the user. (minimum 30 seconds)
```

File: docs/sources/alerting/notifications.md [hunk 12]
```diff
@@ -130,6 +130,23 @@ Move any existing rules using `custom_details.myMetric` to `custom_details.queri
...
+Expire | How many seconds your notification will continue to be retried for (maximum 86400 seconds)
+Alerting sound | The sound for alerting notifications
+OK sound | The sound for OK notifications
```
Comment (oddlittlebird) [inline docs/sources/alerting/notifications.md:149]:
```suggestion
OK sound | The sound for OK notifications

```

**Review by oddlittlebird — CHANGES_REQUESTED** [review]
Hi there! I have a few suggestions to improve the documentation.

**Review by oddlittlebird — APPROVED** [review]
LGTM

**Review by marefr — APPROVED** [review]
LGTM

**Pull request merged**